analytics/location_heatmaps/geo_utils.py

```diff
@@ -217,43 +288,129 @@ def split_regions(tree_prefix_list,
+      neg_prefix = tree_prefix_list[i]
+      pos_prefix = tree_prefix_list[i + 1]
+
+      # check whether the tree has reached the bottom
+      if len(pos_prefix.split('/')) >= image_bit_level:
+        continue
+
+      # total = pos_count + neg_count
+      # p = pos_count / total
+      # confidence = np.sqrt((1-p)*p/total)
+      # error bound propagation.
+      # confidence +/- noise
+      # pos_count/total +/- (confidence+conf_noise) => 95% interval for 95% noise interval.
+
+      if pos_count > threshold and neg_count > threshold:
+        neg_child = get_default_children(positivity, split='neg')
+        pos_child = get_default_children(positivity, split='pos')
+        for j in range(len(pos_child)):
+          new_prefix = f'{neg_prefix}/{neg_child[j]}'
+          if not new_tree.has_key(new_prefix):
+            fresh_expand += 1
+            new_tree[new_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(new_prefix)
+
+            new_prefix = f'{pos_prefix}/{pos_child[j]}'
+            new_tree[new_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(new_prefix)
+      else:
+        if collapse_threshold is not None and \
+            (
+                pos_count < collapse_threshold or neg_count < collapse_threshold) and \
+            len(pos_prefix) > 3 and len(neg_prefix) > 3:
+
+          old_prefix = neg_prefix[:-4]
+          collapsed += 1
+          if not new_tree.has_key(old_prefix):
+            created += 1
+            new_tree[old_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(old_prefix)
+
+            old_prefix = pos_prefix[:-4]
+            new_tree[old_prefix] = len(new_tree_prefix_list)
+            new_tree_prefix_list.append(old_prefix)
+        else:
+          unchanged += 1
+          new_tree[f'{neg_prefix}'] = len(new_tree_prefix_list)
+          new_tree_prefix_list.append(f'{neg_prefix}')
+          new_tree[f'{pos_prefix}'] = len(new_tree_prefix_list)
+          new_tree_prefix_list.append(f'{pos_prefix}')
+  else:
+    for i in range(len(tree_prefix_list)):
+      if expand_all:
+        count = threshold + 1
       else:
-        unchanged += 1
-        new_tree[f'{prefix}'] = len(new_tree_prefix_list)
-        new_tree_prefix_list.append(f'{prefix}')
+        if count_min:
+          count = count_min.query(tree_prefix_list[i])
+        else:
+          count = vector_counts[i]
+      prefix = tree_prefix_list[i]
+
+      # check whether the tree has reached the bottom
+      if len(prefix.split('/')) >= image_bit_level:
+        continue
+      if last_result is not None:
+        (last_prefix, last_prefix_pos) = last_result.tree.longest_prefix(prefix)
+        if last_prefix is None:
+          cond = False
+        else:
+          last_count = last_result.sum_vector[last_prefix_pos]
+          p = (last_count - count) / last_count
+          if p <= 0 or count < 5 or last_count < 5:
```

Comment:
please add a comment about this hard-coded, non-overridable count threshold (i.e. note that we use it because the confidence interval formula below is an approximation that relies on n >= 5)